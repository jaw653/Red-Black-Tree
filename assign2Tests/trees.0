The
" quickity 
     quick 
     "
     brown fox
     jumped over the girl and her lazy, lazy dog.
