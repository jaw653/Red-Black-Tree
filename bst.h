/*
 * Author: Jake Wachs
 * Date: 10/25/2017
 * The University of Alabama
 *
 * This is Dr. Lusth's interface, copied and pasted
 */

#ifndef __BST_INCLUDED__
#define __BST_INCLUDED__

#include <stdio.h>

typedef struct bstnode BSTNODE;

extern void    *getBSTNODE(BSTNODE *n);
extern void    setBSTNODE(BSTNODE *n,void *value);
extern BSTNODE *getBSTNODEleft(BSTNODE *n);
extern void    setBSTNODEleft(BSTNODE *n,BSTNODE *replacement);
extern BSTNODE *getBSTNODEright(BSTNODE *n);
extern void    setBSTNODEright(BSTNODE *n,BSTNODE *replacement);
extern BSTNODE *getBSTNODEparent(BSTNODE *n);
extern void    setBSTNODEparent(BSTNODE *n,BSTNODE *replacement);

typedef struct bst BST;

extern BST *newBST(
    void (*)(FILE *,void *),           //display
    int (*)(void *,void *),            //comparator
    void (*)(BSTNODE *,BSTNODE *));          //swapper
extern void    setBSTroot(BST *t,BSTNODE *replacement);
extern BSTNODE *getBSTroot(BST *t);
extern BSTNODE *insertBST(BST *t,void *value);
extern BSTNODE *findBST(BST *t,void *value);
extern BSTNODE *deleteBST(BST *t,void *value);
extern BSTNODE *swapToLeafBST(BST *t,BSTNODE *node);
extern void    pruneLeafBST(BST *t,BSTNODE *leaf);
extern int     sizeBST(BST *t);
extern void    statisticsBST(FILE *fp,BST *t);
extern void    displayBST(FILE *fp,BST *t);

#endif
