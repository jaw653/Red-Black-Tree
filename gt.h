/* Author: Jake Wachs
 * Date: 10/31/2017
 * The University of Alabama
 */

#ifndef __GT_INCLUDED__
#define __GT_INCLUDED__

#include <stdio.h>

typedef struct gt GT;

extern GT *newGT(
    void (*)(FILE *,void *),           //display
    int (*)(void *,void *));           //comparator
extern void insertGT(GT *,void *);
extern int findGT(GT *,void *);
extern void deleteGT(GT *,void *);
extern int sizeGT(GT *);
extern int wordsGT(GT *);
extern void statisticsGT(FILE *,GT *);
extern void displayGT(FILE *,GT *);

#endif
