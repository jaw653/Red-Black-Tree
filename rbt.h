#ifndef __RBT_INCLUDED__
#define __RBT_INCLUDED__

#include <stdio.h>

typedef struct rbt RBT;

extern RBT *newRBT(
    void (*)(FILE *,void *),           //display
    int (*)(void *,void *));           //comparator
extern void insertRBT(RBT *,void *);
extern int findRBT(RBT *,void *);
extern void deleteRBT(RBT *,void *);
extern int sizeRBT(RBT *);
extern int wordsRBT(RBT *);
extern void statisticsRBT(FILE *,RBT *);
extern void displayRBT(FILE *,RBT *);

#endif
