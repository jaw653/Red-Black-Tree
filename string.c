/* Author: Jake Wachs
 * Implementation of STRING data typedef
 */

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <assert.h>
#include <string.h>
#include "string.h"

struct STRING {
  char *value;
};

STRING *newSTRING(char *str) {
  STRING *ptr = malloc( sizeof(STRING) );
  assert(ptr != 0);
  ptr->value = str;
  return ptr;
}

char *getSTRING(STRING *s) {
  return s->value;
}

char *setSTRING(STRING *s, char *val) {
  char *replacedVal = s->value;
  s->value = val;
  return replacedVal;
}

void displaySTRING(FILE *fp, void *v) {
  fprintf(fp, "%s", getSTRING((STRING *) v));
}

void freeSTRING(STRING *s) {
  free(s);
}

int compareSTRING(void *s1, void *s2) {
  char *rawVal1 = getSTRING(s1);
  char *rawVal2 = getSTRING(s2);

  return strcmp(rawVal1, rawVal2);
}
