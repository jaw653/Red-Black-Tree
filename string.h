/* Author: Jake Wachs
 * Wrapper for STRING data typedef
 */

#ifndef __STRING_INCLUDED__
#define __STRING_INCLUDED__

#include <stdio.h>

typedef struct STRING STRING;

extern STRING *newSTRING(char *);
extern char *getSTRING(STRING *);
extern char *setSTRING(STRING *, char *);
extern void displaySTRING(FILE *, void *);
extern int compareSTRING(void *, void *);
extern void freeSTRING(STRING *);

#endif
