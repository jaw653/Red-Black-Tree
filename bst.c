/*
 * Author: Jake Wachs
 * Date: 10/25/2017
 * The University of Alabama
 *
 * This is the implementation file for the bst.h inteface
 */


#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#include "string.h"
#include "bst.h"
#include "queue.h"

struct bstnode {
  bool isLeftChild;
  void *value;      //points to a gtnode or rbtnode
  struct bstnode *parent;
  struct bstnode *left;
  struct bstnode *right;
};

/******************* Helper function signatures *******************************/
static void displayNODE(BST *, FILE *, BSTNODE *, bool);
static void personalSwapper(BSTNODE *, BSTNODE *);
static BSTNODE *findPredecessor(BSTNODE *);
static BSTNODE *findSuccessor(BSTNODE *);
static BSTNODE *insertHelper(BST *, BSTNODE *, BSTNODE *, void *, bool);
static BSTNODE *findHelper(BSTNODE *, int (*)(void *, void *), void *);
//static BSTNODE *traverseRight(BST *, BSTNODE *, bool);
//static BSTNODE *traverseLeft(BST *, BSTNODE *, bool);
static bool isLeaf(BSTNODE *);
static void displayHelper(FILE *fp, BSTNODE *root, BST *t);
static BSTNODE *copyNODE(BSTNODE *giver);
static int findMinDepthBST(BSTNODE *root);
static int findMaxDepthBST(BSTNODE *root);
//static int min(int, int);


/******************************************************************************/
/***                           BSTNODE Functions                            ***/
/******************************************************************************/
static BSTNODE *newBSTNODE(void *value, BSTNODE *parent) {
  BSTNODE *node;
  if (value == NULL) node = NULL;
  else {
    node = (struct bstnode *)malloc(sizeof(struct bstnode));
    node->value = value;
    node->parent = parent;
    node->left = node->right = NULL;
  }

  return node;
}

void *getBSTNODE(BSTNODE *n) {
  if (!n) return NULL;
  return n->value;
}

void setBSTNODE(BSTNODE *n, void *value) {
  if (n == NULL)
    return;
  n->value = value;
}

BSTNODE *getBSTNODEleft(BSTNODE *n) {
  if (n == NULL) return NULL;
  if (!n->left) return NULL;
  return n->left;
}

void setBSTNODEleft(BSTNODE *n, BSTNODE *replacement) {
  if (n == NULL)
    return;
  n->left = replacement;
/*
  if (n) printf("n is good\n");
  if (n->left) printf("n->left is good\n");
  if (replacement != NULL) {
    n->left->value = replacement->value;
    n->left->parent = replacement->parent;
    n->left->left = replacement->left;
    n->left->right = replacement->right;
  }
  else {
    n = NULL;
  }
*/
}

BSTNODE *getBSTNODEright(BSTNODE *n) {
  if (n == NULL) return NULL;
  if (!n->right) return NULL;
  return n->right;
}

void setBSTNODEright(BSTNODE *n, BSTNODE *replacement) {
  //if (n == NULL) printf("n is null\n");
  if (n == NULL)
    return;
  n->right = replacement;
/*
  if (replacement != NULL) {
    n->right->value = replacement->value;
    n->right->parent = replacement->parent;
    n->right->left = replacement->left;
    n->right->right = replacement->right;
  }
  else {
    n = NULL;
  }
*/
}

BSTNODE *getBSTNODEparent(BSTNODE *n) {
  if (n == NULL) return NULL;
  if (!n->parent) return NULL;
  return n->parent;
}

void setBSTNODEparent(BSTNODE *n, BSTNODE *replacement) {
  if (n == NULL)
    return;
  n->parent = replacement;
/*
  n->parent->value = replacement->value;
  n->parent->parent = replacement->parent;
  n->parent->left = replacement->left;
  n->parent->right = replacement->right;
*/
}


/******************************************************************************/
/***                           BST Functions                                ***/
/******************************************************************************/
struct bst {
  struct bstnode *root;
  int size;
  int numWords;
  void (*display)(FILE *, void *);
  int (*comparator)(void *, void *);
  void (*swapper)(BSTNODE *, BSTNODE *);
};

BST *newBST(
  void (*d)(FILE *, void *),
  int (*c)(void *, void *),
  void (*s)(BSTNODE *, BSTNODE *))
{

    BST *tree = malloc(sizeof(struct bst));
    tree->display = d;
    tree->comparator = c;
    tree->swapper = s;

    tree->root = newBSTNODE(NULL, tree->root);
    tree->size = 0;

    return tree;
}

void setBSTroot(BST *t, BSTNODE *replacement) {
  t->root = replacement;
/*
  t->root->value = replacement->value;
  t->root->parent = replacement->parent;
  t->root->left = replacement->left;
  t->root->right = replacement->right;
*/
}

BSTNODE *getBSTroot(BST *t) {
  if (!t) return NULL;
  if (!t->root) return NULL;

  return t->root;
}

BSTNODE *insertBST(BST *t, void *value) {
  t->numWords += 1;
  t->root = insertHelper(t, t->root, NULL, value, false);
  return t->root;
}

BSTNODE *findBST(BST *t, void *value) {
  if (t->size == 0) {
    return NULL;
  }
  else {
    return findHelper(t->root, t->comparator, value);
  }
}

BSTNODE *deleteBST(BST *t, void *value) {
  if (t->size == 0) {
    return NULL;
  }

  /* Must be root */
  if (t->size == 1) {
    BSTNODE *node1 = findBST(t, value);

    if (node1 == NULL) {
      printf("Value ");
      t->display(stdout, value);
      printf(" not found.\n");
    }

    BSTNODE *returnVal = copyNODE(node1);

    t->root = NULL;
    t->size -= 1;

    return returnVal;
  }

  BSTNODE *node = findBST(t, value);

  if (node == NULL) {
    printf("Value ");
    t->display(stdout, value);
    printf(" not found.\n");
    return NULL;
  }

  node = swapToLeafBST(t, node);

  BSTNODE *returnNODE = copyNODE(node);

  pruneLeafBST(t, node);

  return returnNODE;
}

BSTNODE *swapToLeafBST(BST *t, BSTNODE *node) {
  if (isLeaf(node))
    return node;

  BSTNODE *next = findPredecessor(node);
  if (next == NULL) {
    next = findSuccessor(node);
  }

  if (t->swapper) {
    t->swapper(node, next);
  }
  else {
    personalSwapper(node, next);
  }

  return swapToLeafBST(t, next);
/*
  if (node == NULL)
    return NULL;

  if (t->swapper) {
    if (node->left) {
      t->swapper(node, node->left);
      while (node->right) {
        t->swapper(node, node->right);
        node = node->right;
      }
      node = swapToLeafBST(t, node->left);
    }
    else if (node->right) {
      t->swapper(node, node->right);
      while (node->left) {
        t->swapper(node, node->left);
        node = node
      }
      node = swapToLeafBST(t, node->right);
    }
  }
  else {
    if (node->left) {
      void *tmp = node->value;
      node->value = node->left->value;
      node->left->value = tmp;
      node = swapToLeafBST(t, node->left);
    }
    else if (node->right) {
      void *tmp = node->value;
      node->value = node->right->value;
      node->right->value = tmp;
      node = swapToLeafBST(t, node->right);
    }
  }

  return node;
*/
}

void pruneLeafBST(BST *t, BSTNODE *leaf) {
  if (sizeBST(t) == 1) {
    t->root = NULL;
  }

  BSTNODE *parent = getBSTNODEparent(leaf);

//  if (getBSTNODEleft(parent))
  /* If left child */
  if (t->comparator(getBSTNODE(leaf), getBSTNODE(getBSTNODEleft(parent))   ) == 0) {
    leaf->parent->left = NULL;
  }
  /* If right child */
  else {
    leaf->parent->right = NULL;
  }
  //setBSTNODE(leaf, NULL);
  //setBSTroot(t, newBSTNODE(NULL, NULL));
  //leaf = NULL;
  t->size -= 1;
}

int sizeBST(BST *t) {
  return t->size;
}

void statisticsBST(FILE *fp, BST *t) {
  fprintf(fp, "Nodes: %d\n", t->size);
  fprintf(fp, "Minimum depth: %d\n", findMinDepthBST(t->root));
  fprintf(fp, "Maximum depth: %d\n", findMaxDepthBST(t->root));
}

void displayBST(FILE *fp, BST *t) {
  displayHelper(fp, t->root, t);
}







/******************************************************************************/
/***                           Helper Functions                             ***/
/******************************************************************************/
static void displayNODE(BST *t, FILE *fp, BSTNODE *node, bool isRoot) {
  fprintf(fp, " ");

  if (isLeaf(node))
    fprintf(fp, "=");

  t->display(fp, node->value);
  fprintf(fp, "(");
  t->display(fp, getBSTNODEparent(node)->value);
  fprintf(fp, ")-");

  if (!isRoot) {
    BSTNODE *parent = getBSTNODEparent(node);
    BSTNODE *parLeft = getBSTNODEleft(parent);
    void *nodeVal;
    void *parLeftVal;

    if(node)
      nodeVal = node->value;
    else
      nodeVal = NULL;

    if (parLeft)
      parLeftVal = parLeft->value;
    else
      parLeftVal = NULL;

    if (nodeVal == NULL) {
      if (parLeftVal == NULL) {
        fprintf(fp, "l");
        return;
      }
      else {
        fprintf(fp, "r");
        return;
      }
    }
    else {
      if (parLeftVal == NULL) {
        fprintf(fp, "r");
        return;
      }
    }

    if (t->comparator(nodeVal, parLeftVal) == 0)
      fprintf(fp, "l");
    else
      fprintf(fp, "r");
  }
}

static void personalSwapper(BSTNODE *n1, BSTNODE *n2) {
  void *tmpval = n1->value;
  n1->value = n2->value;
  n2->value = tmpval;
}

static BSTNODE *findPredecessor(BSTNODE *node) {
  if (node == NULL)
    return NULL;
  if (node->left == NULL)
    return NULL;

  node = node->left;

  BSTNODE *predecessor = node;
  while (predecessor->right) {
    predecessor = predecessor->right;
  }

  return predecessor;
}

static BSTNODE *findSuccessor(BSTNODE *node) {
  if (node == NULL)
    return NULL;
  if (node->right == NULL)
    return NULL;

  node = node->right;
  BSTNODE *successor = node;
  while (successor->left) {
    successor = successor->left;
  }

  return successor;
}

static BSTNODE *insertHelper(BST *t, BSTNODE* root, BSTNODE *parent, void *value, bool isLeftChild) {
  if (root == NULL) {
    root = newBSTNODE(value, parent);
    root->isLeftChild = isLeftChild;

    if (parent == NULL)
      root->parent = root;

    t->size += 1;

    return root;
  }
  else if (t->comparator(value, root->value) < 0) {
    root->left = insertHelper(t, root->left, root, value, true);
  }
  else {
    root->right = insertHelper(t, root->right, root, value, false);
  }

  return root;
}

static BSTNODE *findHelper(BSTNODE *root, int (*comparator)(void *, void *), void *value) {
  //if (value->value) printf("FIND VALUE->VALUE IS NOT GOOD\n");
  if (root == NULL)
    return NULL;

  if (comparator(value, root->value) == 0) {
    return root;
  }
  else if (comparator(value, root->value) < 0) {
    return findHelper(getBSTNODEleft(root), comparator, value);
  }
  else {
    return findHelper(getBSTNODEright(root), comparator, value);
  }
}

/*
static BSTNODE *traverseRight(BST *t, BSTNODE *node, bool useSwapper) {
  while (node->right) {
    if (useSwapper == true) {
      t->swapper(node, node->right);
    }
    else {
      //printf("swapping values %lf and %lf\n", getREAL(node->value), getREAL(node->right->value));
      void *tmp = node->value;
      node->value = node->right->value;
      node->right->value = tmp;
    }

    node = node->right;
  }

  return node;
}


static BSTNODE *traverseLeft(BST *t, BSTNODE *node, bool useSwapper) {
  while (node->left) {
    if (useSwapper == true) {
      t->swapper(node, node->left);
    }
    else {
      void *tmp = node->value;
      node->value = node->left->value;
      node->left->value = tmp;
    }

    node = node->left;
  }

  return node;
}
*/
static bool isLeaf(BSTNODE *node) {
  if (node == NULL) return false;
  if (!node->left && !node->right) return true;
  else return false;
}

static void displayHelper(FILE *fp, BSTNODE *root, BST *t) {
  if (t->size == 0) {
    fprintf(fp, "EMPTY");
    return;
  }

  QUEUE *currQueue = newQUEUE(NULL);
  QUEUE *nextQueue = newQUEUE(NULL);

  enqueue(currQueue, root);

  int height = findMaxDepthBST(root);
  int level = 0;
  bool isRoot = true;

  while (level < height) {
    fprintf(fp, "%d:", level);

    while (sizeQUEUE(currQueue) > 0) {
      if (getBSTNODEleft(peekQUEUE(currQueue)))
        enqueue(nextQueue, getBSTNODEleft(peekQUEUE(currQueue)));
      if (getBSTNODEright(peekQUEUE(currQueue)))
        enqueue(nextQueue, getBSTNODEright(peekQUEUE(currQueue)));

      if (isRoot) {
        displayNODE(t, fp, dequeue(currQueue), true);
        isRoot = false;
      }
      else {
        displayNODE(t, fp, dequeue(currQueue), false);
      }
    }
    level += 1;
    if (sizeQUEUE(nextQueue) > 0)
      fprintf(fp, "\n%d:", level);
//printf("size of next queue is: %d\n", sizeQUEUE(nextQueue));

    while (sizeQUEUE(nextQueue) > 0) {
      if (getBSTNODEleft(peekQUEUE(nextQueue)))
        enqueue(currQueue, getBSTNODEleft(peekQUEUE(nextQueue)));
      if (getBSTNODEright(peekQUEUE(nextQueue)))
        enqueue(currQueue, getBSTNODEright(peekQUEUE(nextQueue)));

      displayNODE(t, fp, dequeue(nextQueue), false);
    }

    level += 1;
    if (sizeQUEUE(currQueue) > 0)
      fprintf(fp, "\n");
  }
/******************************************************************************/
/*
  if (t->size == 0) {
    fprintf(fp, "EMPTY");
    return;
  }
  else {
    QUEUE *nodesQueue = newQUEUE(NULL);
    int nodesInCurrLevel = 1;
    int nodesInNextLevel = 0;
    int printLevel = 1;

    enqueue(nodesQueue, root);

    bool isRoot = true;
    int height = findMaxDepthBST(root);

    int i = 0;
    while (sizeQUEUE(nodesQueue) > 0) {
      BSTNODE *currNode = dequeue(nodesQueue);
      nodesInCurrLevel -= 1;

      if (currNode && i <= height) {
        if (isRoot) {
          fprintf(fp, "%d:", 0);
          fprintf(fp, " ");
          displayNODE(t, fp, currNode, true);
          isRoot = false;
        }
        else {
          fprintf(fp, " ");
          displayNODE(t, fp, currNode, false);
        }
      }

      if (getBSTNODEleft(currNode)) {
        enqueue(nodesQueue, getBSTNODEleft(currNode));
        nodesInNextLevel += 1;
      }
      if (getBSTNODEright(currNode)) {
        enqueue(nodesQueue, getBSTNODEright(currNode));
        nodesInNextLevel += 1;
      }
      //  nodesInNextLevel += 2;

      if (nodesInCurrLevel == 0) {
        i += 1;
        if (i == height) break;
        fprintf(fp, "\n");
        fprintf(fp, "%d:", printLevel);

        nodesInCurrLevel = nodesInNextLevel;
        nodesInNextLevel = 0;
        printLevel += 1;
      }
    }
  }
*/
/******************************************************************************/
}

static BSTNODE *copyNODE(BSTNODE *giver) {
  if (giver == NULL) return NULL;

  BSTNODE *newNode = malloc(sizeof(struct bst));

  if (giver->value)
    newNode->value = giver->value;
  else newNode->value = NULL;

  if (giver->parent)
    newNode->parent = giver->parent;
  else newNode->parent = NULL;

  if (giver->left)
    newNode->left = giver->left;
  else newNode->left = NULL;

  if (giver->right)
    newNode->right = giver->right;
  else newNode->right = NULL;

  if (giver->isLeftChild)
    newNode->isLeftChild = giver->isLeftChild;
  else newNode->isLeftChild = NULL;

  return newNode;
}

static int findMinDepthBST(BSTNODE *root) {
  if (root == NULL)
    return 0;

  int Lmin = findMinDepthBST(root->left);
  int Rmin = findMinDepthBST(root->right);

  return (Lmin<Rmin?Lmin:Rmin) + 1;
}

static int findMaxDepthBST(BSTNODE *root) {
  if (root == NULL) return 0;
  //  printf("flag\n");

  int L_depth, R_depth;
  L_depth = findMaxDepthBST(getBSTNODEleft(root));
  R_depth = findMaxDepthBST(getBSTNODEright(root));

  return (L_depth>R_depth?L_depth:R_depth) + 1;
}

/*
static int min(int a, int b) {
  if (a < b) return a;
  else if (a > b) return b;
  else {
    return a;
  }
}
*/
