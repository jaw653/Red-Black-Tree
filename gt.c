/* Author: Jake Wachs
 * Date: 10/31/2017
 * The University of Alabama
 */

#include <stdio.h>
#include <stdlib.h>

#include "bst.h"
#include "gt.h"

struct gt {
  struct bst *tree;
  int totalWords;
  void (*display)(FILE *, void *);
  int (*comparator)(void *, void *);
};

typedef struct GTNODE GTNODE;
struct GTNODE {
  void *value;      //holds the actual value, STRING or INTEGER or whatever
  int frequency;
  void (*display)(FILE *, void *);
  int (*comparator)(void *, void *);
};

static GTNODE *newGTNODE(void *value, void (*)(FILE *, void *), int (*)(void *, void *));
static void displayGTNODE(FILE *, void *);
static void swapGTNODE(BSTNODE *, BSTNODE *);
static int compareGTNODE(void *, void *);
static int findMinDepthGT(BSTNODE *);
static int findMaxDepthGT(BSTNODE *);

GT *newGT(void (*d)(FILE *, void *), int (*c)(void *, void *)) {
  GT *t = malloc(sizeof(struct gt));
  t->tree = newBST(displayGTNODE, compareGTNODE, swapGTNODE);
  t->display = d;
  t->comparator = c;
  t->totalWords = 0;

  return t;
}

void insertGT(GT *t, void *value) {
  GTNODE *valueNode = newGTNODE(value, t->display, t->comparator);

  BSTNODE *valueToFind = findBST(t->tree, valueNode);

  /* If value is in the tree, just increment it */
  if (valueToFind != NULL) {
    GTNODE *nodeToIncrement = getBSTNODE(valueToFind);
    nodeToIncrement->frequency += 1;
  }
  else {
    insertBST(t->tree, valueNode);
  }

  t->totalWords += 1;
//  printf("totalWords is: %d\n", t->totalWords);
}

int findGT(GT *t, void *value) {
  GTNODE *valueNode = newGTNODE(value, t->display, t->comparator);
  GTNODE *p = getBSTNODE(findBST(t->tree, valueNode));

  /* Value is not in the tree */
  if (p == NULL)
    return 0;
  else
    return p->frequency;
}

void deleteGT(GT *t, void *value) {
  GTNODE *valueNode = newGTNODE(value, t->display, t->comparator);
  //printf("value node to find is: %s\n", getSTRING(valueNode->value));
//  printf("delete value is: %s\n", getSTRING(valueNode->value));


  BSTNODE *valueToDelete = findBST(t->tree, valueNode);
  //if (valueToDelete == NULL) printf("value is null\n");

  if (valueToDelete != NULL) {
    GTNODE *gtToDelete = getBSTNODE(valueToDelete);
    gtToDelete->frequency -= 1;
//printf("%s decremented to: %d\n", getSTRING(gtToDelete->value), gtToDelete->frequency);
    if (gtToDelete->frequency <= 0) {
//printf("deleteBST(%s)\n", getSTRING(gtToDelete->value));
      deleteBST(t->tree, valueNode);
    }

    t->totalWords -= 1;
  }
}

int sizeGT(GT *t) {
  return sizeBST(t->tree);
}

int wordsGT(GT *t) {
  return t->totalWords;
}

void statisticsGT(FILE *fp, GT *t) {
  fprintf(fp, "Words/Phrases: %d\n", t->totalWords);
  fprintf(fp, "Nodes: %d\n", sizeBST(t->tree));
  fprintf(fp, "Minimum depth: %d\n", findMinDepthGT(getBSTroot(t->tree)));
  fprintf(fp, "Maximum depth: %d\n", findMaxDepthGT(getBSTroot(t->tree)));
}

void displayGT(FILE *fp, GT *t) {
  displayBST(fp, t->tree);
  fprintf(fp, "\n");
}

/******************************************************************************/
/***                          Helper Functions                              ***/
/******************************************************************************/
static GTNODE *newGTNODE(void *value, void (*d)(FILE *, void *), int (*c)(void *, void*)) {
  GTNODE *node = malloc(sizeof(struct GTNODE));
  node->value = value;
  node->frequency = 1;
  node->display = d;
  node->comparator = c;

  return node;
}

static void displayGTNODE(FILE *fp, void *value) {
  GTNODE *node = value;
  node->display(fp, node->value);
  if (node->frequency > 1) {
    fprintf(fp, "-");
    fprintf(fp, "%d", node->frequency);
  }
}

static void swapGTNODE(BSTNODE *n1, BSTNODE *n2) {
  GTNODE *ra = getBSTNODE(n1);
  GTNODE *rb = getBSTNODE(n2);

  void *vtemp = ra->value;
  ra->value = rb->value;
  rb->value = vtemp;

  int ctemp = ra->frequency;
  ra->frequency = rb->frequency;
  rb->frequency = ctemp;
}

static int compareGTNODE(void *n1, void *n2) {
  GTNODE *node1 = n1;
  GTNODE *node2 = n2;

  if (node1 == NULL) {
    if (node2 == NULL)
      return 0;
    else
      return -1;
  }
  if (node2 == NULL) {
    return 1;
  }

  return node1->comparator(node1->value, node2->value);
}

static int findMinDepthGT(BSTNODE *root) {
  if (root == NULL)
    return 0;

  int Lmin = findMinDepthGT(getBSTNODEleft(root));
  int Rmin = findMinDepthGT(getBSTNODEright(root));

  return (Lmin<Rmin?Lmin:Rmin) + 1;
}

static int findMaxDepthGT(BSTNODE *root) {
  if (root == NULL)
    return 0;

  int Ldepth = findMaxDepthGT(getBSTNODEleft(root));
  int Rdepth = findMaxDepthGT(getBSTNODEright(root));

  return (Ldepth>Rdepth?Ldepth:Rdepth) + 1;
}
