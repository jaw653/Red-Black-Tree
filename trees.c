/* Author: Jake Wachs
 * Date: 11/5/2017
 * The University of Alabama
 */

/*
 * Questions:
            -should i print an error message if there are no cmdln args? or should I do something anyway
            -need to check to handle empty files
 */

#include <stdio.h>
#include <string.h>

#include "bst.h"
#include "rbt.h"
#include "gt.h"

int main(int argc, char *argv[]) {

  /* If no command line arguments */
  if (argc == 1) {
    //print error message?
    return 0;
  }

  char *corpusName, *commandsName;
  FILE *corpus, *commands;
  if (argc == 4) {
    corpusName = argv[argc-2];
    commandsName = argv[argc-1];
  }
  else if (argc == 5) {
    corpusName = argv[argc - 3];
    commandsName = argv[argc-2];
  }

  corpus = fopen(corpusName, "r");
  commands = fopen(commandsName, "r");

  //read info in, populate tree(s)?

  if (corpus) fclose(corpus);
  if (commands) fclose(commands);

  if (strcmp(argv[argc - (argc-1)], "-g") == 0) {
    //green tree opts
  }
  else {
    //red/black tree opts
  }

  return 0;
}
