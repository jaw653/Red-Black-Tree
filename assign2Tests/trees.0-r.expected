0: jumped-B(jumped-B)-
1: fox-R(jumped-B)-l quickity quick-R(jumped-B)-r
2: brown-B(fox-R)-l girl-B(fox-R)-r over-B(quickity quick-R)-l =the-2-B(quickity quick-R)-r
3: =and-R(brown-B)-l =dog-R(brown-B)-r =her-R(girl-B)-r =lazy-2-R(over-B)-l
